firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/crcelements.txt

```diff
@@ -0,0 +1,131 @@
+      THEN("the payload given in the CRCElement constructor is independent of the input body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor")
+
+/* write parse roundtrip
+Scenario: CRCElement correctly preserves data in roundtrip writing and parsing
+
+  GIVEN("A CRC Element initialized with a non empty payload buffer with a capacity of 254 bytes")
+
+    WHEN("A body with crc and payload is generated, parsed and generated again")
+      # write  
+      THEN("the crc accessor method returns 0 before the first write method is called")
+      THEN("The first write method call returns ok status")
+      THEN("After the first write method is called, the crc accessor method returns a value equal to the result of directly computing a CRC on the payload.")
+      THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")
+      THEN("The body's payload section correctly stores the payload as '0x01 0x05 0x01 0x02 0x03 0x04 0x05'")
+      THEN("The output buffer is as expected")
+      # parse
+      THEN("The parse method reports ok status")
+      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
+      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
+      THEN("the payload given in the CRCElement constructor is independent of the input body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor")
+      # write  
+      THEN("The second write method call returns ok status")
+      THEN("After the second write method is called, the crc accessor method returns a value equal to the result of directly computing a CRC on the payload.")
+      THEN("The CRC field of the body's header matches the value returned by the crc accessor method.")
+      THEN("The body's payload section correctly stores the payload as '0x01 0x05 0x01 0x02 0x03 0x04 0x05'")
+      THEN("The output buffer is as expected")
+
+/* CRCElementReceiver
+Scenario: CRCElementReceiver correctly parses CRCElement bodies and performs consistency checking on them
+
+  GIVEN("A CRC element receiver of capacity 254 bytes")
+    WHEN("A body with a complete 4-byte header, and a payload consistent with the CRC field of the header, is given to the receiver")
+      THEN("the transform method reports ok status")
+      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
+      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
+      THEN("the payload buffer updated in the CRCElement is independent of the input body")
+
+    WHEN("A body without a complete 4-byte header and empty payload is given to the crc element receiver")
```

Comment:
I don't understand how this WHEN is different from the next WHEN - I think one of these was the original one and the other was a suggested revision. I would go with something like "WHEN a body with an incomplete header and an empty payload is..." or "WHEN A body without a complete 4-byte header is..." (the latter is a direct copy from tests above, but maybe the tests above should be changed to be more like the former). The current wording, "WHEN A body without a complete 4-byte header and empty payload is...", is a bit ambiguous and could be misinterpreted as meaning "WHEN A body with an incomplete 4-byte header and a non-empty payload is..."

Reply:
Sorry about the ambiguity, so in this test case I'm basically using an empty `input_buffer`, in the latter test we are using a body with crc header less than 4 bytes, so 1,2 and 3 respectively so there it's the case of `incomplete header`, here there's nothing.
so maybe the WHEN should be something like WHEN("An empty input_buffer is given to the crc element receiver")

Reply:
Yes, I like that - good idea to test an empty input as distinct from a non-empty too-small input